python/neuroglancer/pipeline/volumes/cloudvolume.py

```diff
@@ -0,0 +1,486 @@
+from __future__ import print_function
+
+from collections import namedtuple
+import json
+import os
+import re
+
+import numpy as np
+from tqdm import tqdm
+
+import neuroglancer
+from neuroglancer.lib import clamp, xyzrange, Vec, Bbox, min2, max2, find_closest_divisor
+from volumes import Volume, VolumeCutout, generate_slices
+from neuroglancer.pipeline.storage import Storage
+
+__all__ = [ 'CloudVolume', 'EmptyVolumeException' ]
+
+ExtractedPath = namedtuple('ExtractedPath', 
+  ('protocol','bucket_name', 'dataset_name','layer_name')
+)
+
+class EmptyVolumeException(Exception):
+    pass
+
+class CloudVolume(Volume):
+  """
+  CloudVolume represents an interface to a dataset layer at a given
+  mip level. You can use it to send and receive data from neuroglancer
+  datasets on supported hosts like Google Storage, S3, and local Filesystems.
+
+  Required:
+    cloudpath: Path to the dataset layer. This should match storage's supported
+      providers.
+
+      e.g. Google: gs://neuroglancer/$DATASET/$LAYER/
+           S3    : s3://neuroglancer/$DATASET/$LAYER/
+           Lcl FS: file:///tmp/$DATASET/$LAYER/
+  Optional:
+    mip: (int) Which level of downsampling to read to/write from. 0 is the highest resolution.
+    bounded: (bool) If a region outside of volume bounds is accessed:
```

Comment:
- The bounds are specified in the info file in the cloud that gets downloaded.
- I haven't seen any neuroglancer chunks that have negative numbers in them. I have no idea if it would work.
- My interpretation of that would be that it's equivalent to [100, 150], but I'm not sure it's a great idea to mix and match positive and negative numbers.